### PR TITLE
Infer additional item fields in `BlocksBlock` and `ListBlock`

### PR DESCRIPTION
**Pull request by johnnyomair**
Additional fields in the `item` prop of `AdditionalItemContextMenuItems` and `AdditionalItemContent` will be typed correctly if the `additionalItemFields` option is strongly typed.

- [x] Add changeset (if necessary)

### CODE REVIEW

**Comment by johnnyomair** [comment]
Please ignore the ESLint disables and `@ts-expect-error`s for now, I'd just like to get early feedback if this is even the correct approach

File: demo/admin/src/userGroups/userGroupAdditionalItemFields.ts [hunk 1]
```diff
@@ -1,4 +1,7 @@
-const userGroupAdditionalItemFields = {
+import { BlocksBlockAdditionalItemField, ListBlockAdditionalItemField } from "@comet/blocks-admin";
+import { GQLUserGroup } from "@src/graphql.generated";
+
+const userGroupAdditionalItemFields: { userGroup: ListBlockAdditionalItemField<GQLUserGroup> | BlocksBlockAdditionalItemField<GQLUserGroup> } = {
```
Comment (johnnyomair) [inline demo/admin/src/userGroups/userGroupAdditionalItemFields.ts:4]: If `additionalItemFields` in either `createListBlock` or `createBlocksBlock` is typed correctly, the fields will be inferred correctly.

**Comment by johnnyomair** [comment]
Another approach would be to explicitly use the generated block types for the factories, for instance `const PageContentBlock = createBlocksBlock<PageContentBlockData, PageContentBlockInput>(...)`. We haven't used this approach anywhere yet, so I decided to go with the "infer types from options" approach instead.

File: packages/admin/blocks-admin/src/blocks/factories/createBlocksBlock.tsx [hunk 2]
```diff
@@ -588,13 +624,12 @@ export function createBlocksBlock({
...
-                                                                            const blockToMove = blocks.splice(from, 1)[0];
-                                                                            blocks.splice(to, 0, blockToMove);
-                                                                            return { ...prevState, blocks };
-                                                                        });
+                                                                    moveBlock={(dragIndex: number, hoverIndex: number) => {
+                                                                        const blocks = [...state.blocks];
+                                                                        const dragItem = state.blocks[dragIndex];
+                                                                        blocks[dragIndex] = state.blocks[hoverIndex];
+                                                                        blocks[hoverIndex] = dragItem;
+                                                                        updateState((prevState) => ({ ...prevState, blocks }));
```
Comment (thomasdax98) [inline packages/admin/blocks-admin/src/blocks/factories/createBlocksBlock.tsx:632]: Why are you doing this? Looks like you are reverting the fix you made in https://github.com/vivid-planet/comet/pull/1344/files?

Comment (johnnyomair) [reply packages/admin/blocks-admin/src/blocks/factories/createBlocksBlock.tsx]: This wasn't intentional. Used an old local copy that was obviously started before the fix. Reverted: b05dc94095878cb7211e2b0c6df7dfeca7523be6

**Comment by johnnyomair** [comment]
> I don't think there's a better way to do this when using generics, as anything may be passed to the function

I tried using a type guard and a conditional function return type, but that didn't work. I think it's okay with to live with the type casts...

**Comment by johnnyomair** [comment]
> Another approach would be to explicitly use the generated block types for the factories, for instance `const PageContentBlock = createBlocksBlock<PageContentBlockData, PageContentBlockInput>(...)`. We haven't used this approach anywhere yet, so I decided to go with the "infer types from options" approach instead.

We may type the block factories correctly for this to work in the future, but this is out of scope of this PR.

**Comment by johnnyomair** [comment]
@nsams @thomasdax98 this PR is ready for review now.

File: .changeset/silly-lemons-compete.md [hunk 3]
```diff
@@ -0,0 +1,7 @@
+---
+"@comet/blocks-admin": patch
```
Comment (johnnyomair) [inline .changeset/silly-lemons-compete.md:2]: Do you think this is a patch or a minor worthy change?

**Comment by johnnyomair** [comment]
@nsams @thomasdax98 please review. Thanks!

File: demo/admin/src/userGroups/userGroupAdditionalItemFields.ts (hunk 1 above)
Comment (nsams) [reply demo/admin/src/userGroups/userGroupAdditionalItemFields.ts:4]: wouldn't it be possible to infer this type?

Comment (nsams) [reply demo/admin/src/userGroups/userGroupAdditionalItemFields.ts:4]: I mean to omit the type here

Comment (johnnyomair) [reply demo/admin/src/userGroups/userGroupAdditionalItemFields.ts:4]: It is possible. However, `userGroup` is then inferred as string, and not as `GQLUserGroup`:

```
Type 'BlocksBlockItem<BlockInterface<any, any, any, any>, { userGroup: string; }>' is not assignable to type 'UserGroupItem'.
  Types of property 'userGroup' are incompatible.
    Type 'string' is not assignable to type 'GQLUserGroup'.ts(2322)
```

Comment (nsams) [reply demo/admin/src/userGroups/userGroupAdditionalItemFields.ts:4]: you could possibly use `"All" as GQLUserGroup`?

**Review by thomasdax98 — DISMISSED** [review]

File: .changeset/silly-lemons-compete.md (hunk 3 above)
Comment (thomasdax98) [reply .changeset/silly-lemons-compete.md:2]: I think patch is okay

File: demo/admin/src/userGroups/userGroupAdditionalItemFields.ts (hunk 1 above)
Comment (johnnyomair) [reply demo/admin/src/userGroups/userGroupAdditionalItemFields.ts:4]: Do you think an as cast is better than the explicit typing?

Comment (nsams) [reply demo/admin/src/userGroups/userGroupAdditionalItemFields.ts:4]: Generally of course not.

But here, if you convert a generic string to an strings-union (don't know how this type is called) the type gets more accurate and can be inferred.

And inferring is in this case better than explicitly defining imho. Especially this `ListBlockAdditionalItemField<GQLUserGroup> | BlocksBlockAdditionalItemField<GQLUserGroup>` is not very nice.

Comment (johnnyomair) [reply demo/admin/src/userGroups/userGroupAdditionalItemFields.ts]: Now I understand! Fixed by b5dc8af4bd337de1694d171229e4422f730030f8

> ListBlockAdditionalItemField<GQLUserGroup> | BlocksBlockAdditionalItemField<GQLUserGroup>

Should I remove those two types from the public API as well?

Comment (nsams) [reply demo/admin/src/userGroups/userGroupAdditionalItemFields.ts]: > Should I remove those two types from the public API as well?

if we don't need it anymore, better remove it

**Review by nsams — APPROVED** [review]

**Pull request merged**